### PR TITLE
Add File::Delete and File::DeleteDir warning flags

### PR DESCRIPTION
**Pull request by Dentomologist**
Adds a flag to File::Delete and File::DeleteDir dictating whether they emit console warnings when the specified path doesn't exist.  The flags are true by default to maintain existing behavior.

Also modifies the return value of DeleteDir when attempting to delete a nonexistent path to better reflect the intention of the function (nothing's there after the call).  This change doesn't break anything because none of the existing callers actually check the return value.

Adds tests for Delete and DeleteDir.

### CODE REVIEW

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by lioncash** [comment]
I don't think a bool is particularly useful visually here, because

`File::Delete(path, false);` doesn't really communicate intent particularly well. Maybe consider using an enum class instead.

**Comment by Dentomologist** [comment]
> I don't think a bool is particularly useful visually here, because
> 
> `File::Delete(path, false);` doesn't really communicate intent particularly well. Maybe consider using an enum class instead.

I considered that but was reluctant to add an enum that was only used by two function parameters.  If you feel that's a better route I can go ahead and do that.

**Comment by lioncash** [comment]
Please do. It's much better to be noisily informed about behavior/intent, than not being informed about it at all.

**Comment by Dentomologist** [comment]
Noted!  I've added the enum class as requested.

File: Source/Core/Common/FileUtil.h [hunk 1]
```diff
@@ -136,12 +136,18 @@ bool CreateDir(const std::string& filename);
 // Creates the full path of fullPath returns true on success
 bool CreateFullPath(const std::string& fullPath);
 
+enum class IfAbsent
```
Comment (leoetlino) [inline Source/Core/Common/FileUtil.h:139]: IfAbsentBehavior perhaps? That's a bit more verbose but I don't think IfAbsent is clear enough; it doesn't sound like the name of an enum class.

Comment (Dentomologist) [reply Source/Core/Common/FileUtil.h]: Done.

File: Source/UnitTests/FileUtil.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+#include <gtest/gtest.h>
+
+#include "Common/FileUtil.h"
+
+class FileUtilTest : public testing::Test
+{
+protected:
+  FileUtilTest()
+      : m_parent_directory{File::CreateTempDir()}, file_path(m_parent_directory + "/file.txt"),
+        directory_path(m_parent_directory + "/dir"),
```
Comment (leoetlino) [inline Source/UnitTests/FileUtil.cpp:14]: This seems dangerous when m_parent_directory can be the empty string (if a temporary directory fails to be created for whatever reason).

Comment (Dentomologist) [reply Source/UnitTests/FileUtil.cpp]: Fixed, thank you for pointing that out.  

There are a few other places in the codebase with a similar problem, but that's an issue for another PR.

**Comment by Dentomologist** [comment]
I refactored a few things in the unit tests so that could use another look.  In addition to renaming the path variables to use the m_ prefix I also separated assertions into separate functions.  This is intended make it clearer exactly what is being tested, and also that the results of the delete shouldn't depend on the value of the IfAbsentBehavior.

File: Source/UnitTests/FileUtil.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+{
+protected:
+  FileUtilTest()
+      : m_parent_directory(File::CreateTempDir()), m_file_path(m_parent_directory + "/file.txt"),
+        m_directory_path(m_parent_directory + "/dir"),
+        m_invalid_path(m_parent_directory + "/invalid.txt")
+  {
+  }
+
+  virtual ~FileUtilTest()
```
Comment (leoetlino) [inline Source/UnitTests/FileUtil.cpp:20]:
```suggestion
  ~FileUtilTest() override
```

File: Source/UnitTests/FileUtil.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      File::IfAbsentBehavior::NoConsoleWarning,
+  };
+
+  const std::string m_parent_directory;
+  const std::string m_file_path;
+  const std::string m_directory_path;
+  const std::string m_invalid_path;
+};
+
+void DeleteShouldNotRemoveDirectory(const std::string& path, File::IfAbsentBehavior behavior)
```
Comment (leoetlino) [inline Source/UnitTests/FileUtil.cpp:47]: This function (and the ones below) should be marked static.

File: Source/UnitTests/FileUtil.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  FileUtilTest()
+      : m_parent_directory(File::CreateTempDir()), m_file_path(m_parent_directory + "/file.txt"),
+        m_directory_path(m_parent_directory + "/dir"),
+        m_invalid_path(m_parent_directory + "/invalid.txt")
+  {
+  }
+
+  virtual ~FileUtilTest()
+  {
+    if (m_parent_directory != "")
```
Comment (leoetlino) [inline Source/UnitTests/FileUtil.cpp:22]:
```suggestion
    if (!m_parent_directory.empty())
```

File: Source/UnitTests/FileUtil.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    {
+      File::DeleteDirRecursively(m_parent_directory);
+    }
+  }
+
+  void SetUp()
+  {
+    if (m_parent_directory == "")
```
Comment (leoetlino) [inline Source/UnitTests/FileUtil.cpp:30]:
```suggestion
    if (m_parent_directory.empty())
```

**Review by leoetlino — CHANGES_REQUESTED** [review]

**Comment by Dentomologist** [comment]
Added suggested changes by @leoetlino

**Review by leoetlino — APPROVED** [review]

**Pull request merged**